File: clients/ios-demo/frida/relay_client.js
'use strict';

const dispatchSemaphoreCreate = new NativeFunction(
  Module.getGlobalExportByName('dispatch_semaphore_create'),
  'pointer',
  ['long']
);
const dispatchSemaphoreSignal = new NativeFunction(
  Module.getGlobalExportByName('dispatch_semaphore_signal'),
  'long',
  ['pointer']
);
const dispatchSemaphoreWait = new NativeFunction(
  Module.getGlobalExportByName('dispatch_semaphore_wait'),
  'long',
  ['pointer', 'uint64']
);

const DISPATCH_TIME_FOREVER = uint64('0xffffffffffffffff');

function waitSemaphore(sem) {
  dispatchSemaphoreWait(sem, DISPATCH_TIME_FOREVER);
}

function withSemaphore(block) {
  const sem = dispatchSemaphoreCreate(0);
  block(sem);
  waitSemaphore(sem);
}

function nsString(value) {
  return ObjC.classes.NSString.stringWithString_(value);
}

function jsonStringify(value) {
  const NSJSONSerialization = ObjC.classes.NSJSONSerialization;
  const NSString = ObjC.classes.NSString;
  const nsValue = plainToNSDictionary(value);
  const data = NSJSONSerialization.dataWithJSONObject_options_error_(nsValue, 0, NULL);
  if (data.isNull()) {
    return '{}';
  }
  return NSString.alloc().initWithData_encoding_(data, 4).toString();
}

function jsonParse(text) {
  const NSJSONSerialization = ObjC.classes.NSJSONSerialization;
  const NSString = ObjC.classes.NSString;
  const data = NSString.stringWithString_(text).dataUsingEncoding_(4);
  const obj = NSJSONSerialization.JSONObjectWithData_options_error_(data, 0, NULL);
  if (obj.isNull()) {
    return null;
  }
  return nsDictionaryToPlain(obj);
}

function nsDictionaryToPlain(obj) {
  if (obj.isKindOfClass_(ObjC.classes.NSString)) {
    return obj.toString();
  }
  if (obj.isKindOfClass_(ObjC.classes.NSNumber)) {
    return obj.doubleValue();
  }
  if (obj.isKindOfClass_(ObjC.classes.NSDictionary)) {
    const keys = obj.allKeys();
    const count = keys.count().valueOf();
    const out = {};
    for (let i = 0; i < count; i++) {
      const key = keys.objectAtIndex_(i).toString();
      out[key] = nsDictionaryToPlain(obj.objectForKey_(keys.objectAtIndex_(i)));
    }
    return out;
  }
  if (obj.isKindOfClass_(ObjC.classes.NSArray)) {
    const count = obj.count().valueOf();
    const out = [];
    for (let i = 0; i < count; i++) {
      out.push(nsDictionaryToPlain(obj.objectAtIndex_(i)));
    }
    return out;
  }
  return obj.toString();
}

function plainToNSDictionary(value) {
  const NSDictionary = ObjC.classes.NSDictionary;
  const NSNumber = ObjC.classes.NSNumber;
  if (value === null || value === undefined) {
    return NSDictionary.dictionary();
  }
  if (typeof value === 'string') {
    return nsString(value);
  }
  if (typeof value === 'number') {
    return NSNumber.numberWithDouble_(value);
  }
  if (typeof value === 'boolean') {
    return NSNumber.numberWithBool_(value ? 1 : 0);
  }
  if (Array.isArray(value)) {
    const NSArray = ObjC.classes.NSArray;
    const items = value.map(plainToNSDictionary);
    return NSArray.arrayWithArray_(items);
  }
  if (typeof value === 'object') {
    const keys = Object.keys(value).map(k => nsString(k));
    const vals = Object.keys(value).map(k => plainToNSDictionary(value[k]));
    return NSDictionary.dictionaryWithObjects_forKeys_(vals, keys);
  }
  return nsString(String(value));
}

function urlEncode(value) {
  let out = '';
  const bytes = Memory.allocUtf8String(value);
  let cursor = bytes;
  while (true) {
    const c = cursor.readU8();
    if (c === 0) {
      break;
    }
    const safe = (c >= 0x61 && c <= 0x7a) || (c >= 0x41 && c <= 0x5a) || (c >= 0x30 && c <= 0x39)
      || c === 0x2d || c === 0x5f || c === 0x2e || c === 0x7e;
    if (safe) {
      out += String.fromCharCode(c);
    } else {
      out += '%' + c.toString(16).toUpperCase().padStart(2, '0');
    }
    cursor = cursor.add(1);
  }
  return out;
}

class RelayClient {
  constructor(baseUrl, username, password, clientId, group, platform) {
    this.baseUrl = normalizeBaseUrl(baseUrl);
    this.username = username;
    this.password = password;
    this.clientId = clientId;
    this.group = group;
    this.platform = platform || 'ios';
    this.token = '';
    this.wsUrl = '';
    this.handlers = {};
    this.running = false;
    this.session = ObjC.classes.NSURLSession.sharedSession();
    this.webSocketTask = null;
    this.heartbeatTimer = null;
  }

  registerHandler(action, handler) {
    this.handlers[action] = handler;
  }

  start() {
    if (this.running) {
      return;
    }
    this.running = true;
    setImmediate(() => this.loopForever());
  }

  stop() {
    this.running = false;
    this.stopHeartbeat();
    if (this.webSocketTask) {
      this.webSocketTask.cancelWithCloseCode_reason_(1000, nsString(''));
      this.webSocketTask = null;
    }
  }

  loopForever() {
    let retryAttempt = 0;
    const tick = () => {
      if (!this.running) {
        return;
      }
      let connectedAt = 0;
      Promise.resolve()
        .then(() => {
          this.ensureLoggedIn();
          connectedAt = Date.now();
          return this.connectAndRun();
        })
        .then(() => {
          if (!this.running) {
            return;
          }
          if (Date.now() - connectedAt >= 60000) {
            retryAttempt = 0;
          }
        })
        .catch((err) => {
          if (!this.running) {
            return;
          }
          const message = String(err);
          if (message.includes('401') || message.toLowerCase().includes('unauthorized')) {
            this.token = '';
            this.wsUrl = '';
          }
          if (connectedAt > 0 && Date.now() - connectedAt >= 60000) {
            retryAttempt = 0;
          }
          console.log(`[R0RPC] reconnect scheduled, attempt=${retryAttempt + 1}, reason=${message}`);
        })
        .finally(() => {
          if (!this.running) {
            return;
          }
          const delayMs = retryDelayForAttempt(retryAttempt);
          retryAttempt += 1;
          setTimeout(tick, delayMs);
        });
    };
    tick();
  }

  ensureLoggedIn() {
    if (this.token) {
      return;
    }
    this.login();
  }

  login() {
    const url = ObjC.classes.NSURL.URLWithString_(nsString(`${this.baseUrl}/api/client/login`));
    const request = ObjC.classes.NSMutableURLRequest.requestWithURL_(url);
    request.setHTTPMethod_(nsString('POST'));
    request.setValue_forHTTPHeaderField_(nsString('application/json; charset=UTF-8'), nsString('Content-Type'));

    const body = plainToNSDictionary({
      username: this.username,
      password: this.password,
      clientId: this.clientId,
      group: this.group,
      platform: this.platform,
      maxInFlight: 64
    });
    const bodyData = ObjC.classes.NSJSONSerialization.dataWithJSONObject_options_error_(body, 0, NULL);
    request.setHTTPBody_(bodyData);

    let responseData = null;
    let statusCode = 0;
    let requestError = null;

    withSemaphore((sem) => {
      const completion = new ObjC.Block({
        retType: 'void',
        argTypes: ['object', 'object', 'object'],
        implementation(data, response, error) {
          responseData = data;
          if (response && response.isKindOfClass_(ObjC.classes.NSHTTPURLResponse)) {
            statusCode = response.statusCode();
          }
          requestError = error;
          dispatchSemaphoreSignal(sem);
        }
      });
      this.session.dataTaskWithRequest_completionHandler_(request, completion).resume();
    });

    if (requestError && !requestError.isNull()) {
      throw new Error(requestError.localizedDescription().toString());
    }
    if (statusCode >= 400) {
      throw new Error(`HTTP ${statusCode}`);
    }

    const text = ObjC.classes.NSString.alloc().initWithData_encoding_(responseData, 4).toString();
    const parsed = jsonParse(text);
    if (!parsed || !parsed.token) {
      throw new Error('login succeeded but token is missing');
    }
    this.token = parsed.token;
    this.wsUrl = parsed.wsUrl || this.buildWsUrl();
  }

  buildWsUrl() {
    let wsBase = this.baseUrl;
    if (wsBase.startsWith('https://')) {
      wsBase = `wss://${wsBase.substring(8)}`;
    } else if (wsBase.startsWith('http://')) {
      wsBase = `ws://${wsBase.substring(7)}`;
    }
    return `${wsBase}/api/client/ws?token=${urlEncode(this.token)}`;
  }

  connectAndRun() {
    const wsUrl = this.wsUrl || this.buildWsUrl();
    const url = ObjC.classes.NSURL.URLWithString_(nsString(wsUrl));
    this.webSocketTask = this.session.webSocketTaskWithURL_(url);
    this.webSocketTask.resume();
    this.startHeartbeat();

    return new Promise((resolve, reject) => {
      const readLoop = () => {
        if (!this.running || !this.webSocketTask) {
          resolve();
          return;
        }

        const completion = new ObjC.Block({
          retType: 'void',
          argTypes: ['object', 'object'],
          implementation: (message, error) => {
            if (!this.running) {
              resolve();
              return;
            }
            if (error && !error.isNull()) {
              this.stopHeartbeat();
              this.webSocketTask = null;
              reject(new Error(error.localizedDescription().toString()));
              return;
            }
            let text = '';
            if (message.type() === 0) {
              text = message.string().toString();
            } else if (message.type() === 1) {
              text = ObjC.classes.NSString.alloc().initWithData_encoding_(message.data(), 4).toString();
            }
            if (text) {
              try {
                this.handleIncomingText(text);
              } catch (err) {
                console.log(`[R0RPC] handle message failed: ${err}`);
              }
            }
            readLoop();
          }
        });

        this.webSocketTask.receiveMessageWithCompletionHandler_(completion);
      };

      readLoop();
    });
  }

  handleIncomingText(text) {
    const message = jsonParse(text);
    if (!message || message.type !== 'job' || !message.job) {
      return;
    }
    const job = message.job;
    const action = String(job.action || '');
    const handler = this.handlers[action];
    const startedAt = Date.now();
    const requestId = String(job.requestId || '');

    const respond = (status, httpCode, payload, error) => {
      const latencyMs = Date.now() - startedAt;
      this.sendResult(requestId, status, httpCode, payload || {}, error || '', latencyMs);
    };

    if (!handler) {
      respond('error', 500, {}, `No handler registered for action: ${action}`);
      return;
    }

    try {
      handler(job, respond);
    } catch (err) {
      respond('error', 500, {}, String(err));
    }
  }

  sendResult(requestId, status, httpCode, payload, error, latencyMs) {
    if (!this.webSocketTask) {
      return;
    }
    const envelope = {
      type: 'result',
      result: {
        requestId,
        status,
        httpCode,
        payload,
        error,
        latencyMs
      }
    };
    const jsonText = jsonStringify(envelope);
    const message = ObjC.classes.NSURLSessionWebSocketMessage.messageWithString_(nsString(jsonText));
    this.webSocketTask.sendMessage_completionHandler_(message, new ObjC.Block({
      retType: 'void',
      argTypes: ['object'],
      implementation(sendError) {
        if (sendError && !sendError.isNull()) {
          console.log(`[R0RPC] send result failed: ${sendError.localizedDescription()}`);
        }
      }
    }));
  }

  startHeartbeat() {
    this.stopHeartbeat();
    this.heartbeatTimer = setInterval(() => {
      if (!this.webSocketTask) {
        return;
      }
      const message = ObjC.classes.NSURLSessionWebSocketMessage.messageWithString_(nsString('{"type":"heartbeat"}'));
      this.webSocketTask.sendMessage_completionHandler_(message, NULL);
    }, 5000 + Math.floor(Math.random() * 1500));
  }

  stopHeartbeat() {
    if (this.heartbeatTimer) {
      clearInterval(this.heartbeatTimer);
      this.heartbeatTimer = null;
    }
  }
}

function normalizeBaseUrl(value) {
  let normalized = String(value || '').trim();
  if (!normalized) {
    throw new Error('baseUrl can not be empty');
  }
  if (!normalized.startsWith('http://') && !normalized.startsWith('https://')) {
    normalized = `http://${normalized}`;
  }
  while (normalized.endsWith('/')) {
    normalized = normalized.slice(0, -1);
  }
  return normalized;
}

function retryDelayForAttempt(attempt) {
  const safeAttempt = Math.max(0, Math.min(attempt, 6));
  const capped = Math.min(1000 * (2 ** safeAttempt), 30000);
  if (capped <= 1000) {
    return 1000;
  }
  return 1000 + Math.floor(Math.random() * (capped - 1000 + 1));
}

module.exports = {
  RelayClient
};

File: clients/ios-demo/ios-r0rpc/src/Core/RelayClient.m
#import "RelayClient.h"

static NSString *const kRelayLogTag = @"R0RPC";

static NSTimeInterval const kBaseRetryDelay = 1.0;
static NSTimeInterval const kMaxRetryDelay = 30.0;
static NSTimeInterval const kHeartbeatInterval = 5.0;
static NSTimeInterval const kHeartbeatJitter = 1.5;

@interface RelayClient ()
@property (nonatomic, copy) NSString *baseURL;
@property (nonatomic, copy) NSString *username;
@property (nonatomic, copy) NSString *password;
@property (nonatomic, copy) NSString *clientId;
@property (nonatomic, copy) NSString *group;
@property (nonatomic, copy) NSString *platform;
@property (nonatomic, copy) NSString *token;
@property (nonatomic, copy) NSString *wsURL;
@property (nonatomic, strong) NSMutableDictionary<NSString *, RelayJobHandler> *handlers;
@property (nonatomic, strong) dispatch_queue_t workerQueue;
@property (nonatomic, strong) dispatch_queue_t jobQueue;
@property (nonatomic, strong) NSURLSession *session;
@property (nonatomic, strong) NSURLSessionWebSocketTask *webSocketTask;
@property (nonatomic, strong) dispatch_source_t heartbeatTimer;
@property (nonatomic, assign) BOOL running;
@property (nonatomic, assign) BOOL started;
@property (nonatomic, assign) NSInteger maxInFlight;
@end

@implementation RelayClient

- (instancetype)initWithBaseURL:(NSString *)baseURL
                       username:(NSString *)username
                       password:(NSString *)password
                       clientId:(NSString *)clientId
                          group:(NSString *)group
                       platform:(NSString *)platform {
    self = [super init];
    if (self) {
        _baseURL = [self normalizedBaseURL:baseURL];
        _username = [username copy];
        _password = [password copy];
        _clientId = [clientId copy];
        _group = [group copy];
        _platform = platform.length > 0 ? [platform copy] : @"ios";
        _handlers = [NSMutableDictionary dictionary];
        _workerQueue = dispatch_queue_create("com.r0rpc.relay.worker", DISPATCH_QUEUE_SERIAL);
        _jobQueue = dispatch_queue_create("com.r0rpc.relay.jobs", DISPATCH_QUEUE_CONCURRENT);
        _session = [NSURLSession sessionWithConfiguration:NSURLSessionConfiguration.defaultSessionConfiguration];
        _maxInFlight = 64;
    }
    return self;
}

- (void)registerHandler:(NSString *)action handler:(RelayJobHandler)handler {
    if (action.length == 0 || handler == nil) {
        return;
    }
    self.handlers[action] = [handler copy];
}

- (void)start {
    if (self.started) {
        return;
    }
    self.started = YES;
    self.running = YES;
    dispatch_async(self.workerQueue, ^{
        [self loopForever];
    });
}

- (void)stop {
    self.running = NO;
    [self stopHeartbeat];
    [self.webSocketTask cancelWithCloseCode:NSURLSessionWebSocketCloseCodeNormalClosure reason:nil];
    self.webSocketTask = nil;
}

#pragma mark - Connection loop

- (void)loopForever {
    NSInteger retryAttempt = 0;
    while (self.running) {
        @autoreleasepool {
            NSDate *connectedAt = nil;
            @try {
                [self ensureLoggedIn];
                connectedAt = [NSDate date];
                [self connectAndRun];
                if (!self.running) {
                    return;
                }
                if (connectedAt && [[NSDate date] timeIntervalSinceDate:connectedAt] >= 60.0) {
                    retryAttempt = 0;
                }
            } @catch (NSException *exception) {
                if (!self.running) {
                    return;
                }
                NSString *message = exception.reason ?: exception.name;
                if ([message containsString:@"401"] || [message localizedCaseInsensitiveContainsString:@"unauthorized"]) {
                    self.token = nil;
                    self.wsURL = nil;
                }
                if (connectedAt && [[NSDate date] timeIntervalSinceDate:connectedAt] >= 60.0) {
                    retryAttempt = 0;
                }
                NSLog(@"[%@] reconnect scheduled, attempt=%ld, reason=%@", kRelayLogTag, (long)retryAttempt + 1, message);
            }
            NSTimeInterval delay = [self retryDelayForAttempt:retryAttempt];
            retryAttempt += 1;
            [NSThread sleepForTimeInterval:delay];
        }
    }
}

- (void)connectAndRun {
    NSString *wsURLString = self.wsURL.length > 0 ? self.wsURL : [self buildWSURL];
    NSURL *url = [NSURL URLWithString:wsURLString];
    if (url == nil) {
        @throw [NSException exceptionWithName:@"RelayClientError" reason:@"invalid ws url" userInfo:nil];
    }

    self.webSocketTask = [self.session webSocketTaskWithURL:url];
    [self.webSocketTask resume];
    [self startHeartbeat];

    while (self.running) {
        dispatch_semaphore_t sem = dispatch_semaphore_create(0);
        __block NSString *text = nil;
        __block NSError *readError = nil;

        [self.webSocketTask receiveMessageWithCompletionHandler:^(NSURLSessionWebSocketMessage *message, NSError *error) {
            readError = error;
            if (message.type == NSURLSessionWebSocketMessageTypeString) {
                text = message.string;
            } else if (message.type == NSURLSessionWebSocketMessageTypeData && message.data != nil) {
                text = [[NSString alloc] initWithData:message.data encoding:NSUTF8StringEncoding];
            }
            dispatch_semaphore_signal(sem);
        }];
        dispatch_semaphore_wait(sem, DISPATCH_TIME_FOREVER);

        if (readError != nil || text.length == 0) {
            [self stopHeartbeat];
            [self.webSocketTask cancelWithCloseCode:NSURLSessionWebSocketCloseCodeGoingAway reason:nil];
            self.webSocketTask = nil;
            if (readError != nil) {
                @throw [NSException exceptionWithName:@"RelayClientError" reason:readError.localizedDescription userInfo:nil];
            }
            return;
        }

        [self handleIncomingText:text];
    }
}

#pragma mark - Message handling

- (void)handleIncomingText:(NSString *)text {
    NSData *data = [text dataUsingEncoding:NSUTF8StringEncoding];
    id parsed = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
    if (![parsed isKindOfClass:[NSDictionary class]]) {
        return;
    }
    NSDictionary *message = (NSDictionary *)parsed;
    if (![message[@"type"] isEqualToString:@"job"]) {
        return;
    }
    id jobObject = message[@"job"];
    if (![jobObject isKindOfClass:[NSDictionary class]]) {
        return;
    }
    NSDictionary *job = [(NSDictionary *)jobObject copy];
    dispatch_async(self.jobQueue, ^{
        [self dispatchJob:job];
    });
}

- (void)dispatchJob:(NSDictionary *)job {
    NSString *action = [self stringValue:job[@"action"]];
    RelayJobHandler handler = self.handlers[action];
    NSDate *startedAt = [NSDate date];
    NSString *requestId = [self stringValue:job[@"requestId"]];

    void (^response)(NSString *, NSInteger, NSDictionary *, NSString *) = ^(NSString *status, NSInteger httpCode, NSDictionary *payload, NSString *error) {
        NSTimeInterval latencyMs = [[NSDate date] timeIntervalSinceDate:startedAt] * 1000.0;
        [self sendResultWithRequestId:requestId status:status httpCode:httpCode payload:payload ?: @{} error:error ?: @"" latencyMs:(NSInteger)latencyMs];
    };

    if (handler == nil) {
        response(@"error", 500, @{}, [NSString stringWithFormat:@"No handler registered for action: %@", action]);
        return;
    }

    @try {
        handler(job, response);
    } @catch (NSException *exception) {
        response(@"error", 500, @{}, exception.reason ?: exception.name);
    }
}

- (void)sendResultWithRequestId:(NSString *)requestId
                         status:(NSString *)status
                       httpCode:(NSInteger)httpCode
                        payload:(NSDictionary *)payload
                          error:(NSString *)error
                      latencyMs:(NSInteger)latencyMs {
    NSDictionary *resultBody = @{
        @"requestId": requestId ?: @"",
        @"status": status ?: @"error",
        @"httpCode": @(httpCode),
        @"payload": payload ?: @{},
        @"error": error ?: @"",
        @"latencyMs": @(latencyMs)
    };
    NSDictionary *envelope = @{
        @"type": @"result",
        @"result": resultBody
    };
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:envelope options:0 error:nil];
    NSString *jsonText = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    if (jsonText.length == 0) {
        return;
    }

    NSURLSessionWebSocketMessage *message = [[NSURLSessionWebSocketMessage alloc] initWithString:jsonText];
    dispatch_semaphore_t sem = dispatch_semaphore_create(0);
    [self.webSocketTask sendMessage:message completionHandler:^(NSError *sendError) {
        if (sendError != nil) {
            NSLog(@"[%@] send result failed: %@", kRelayLogTag, sendError.localizedDescription);
        }
        dispatch_semaphore_signal(sem);
    }];
    dispatch_semaphore_wait(sem, DISPATCH_TIME_FOREVER);
}

#pragma mark - Heartbeat

- (void)startHeartbeat {
    [self stopHeartbeat];
    dispatch_queue_t queue = dispatch_get_global_queue(QOS_CLASS_UTILITY, 0);
    self.heartbeatTimer = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, queue);
    if (self.heartbeatTimer == nil) {
        return;
    }
    uint64_t intervalNs = (uint64_t)((kHeartbeatInterval + (arc4random_uniform((uint32_t)(kHeartbeatJitter * 1000)) / 1000.0)) * NSEC_PER_SEC);
    dispatch_source_set_timer(self.heartbeatTimer, dispatch_time(DISPATCH_TIME_NOW, intervalNs), intervalNs, (uint64_t)(0.2 * NSEC_PER_SEC));
    __weak typeof(self) weakSelf = self;
    dispatch_source_set_event_handler(self.heartbeatTimer, ^{
        [weakSelf sendHeartbeat];
    });
    dispatch_resume(self.heartbeatTimer);
}

- (void)stopHeartbeat {
    if (self.heartbeatTimer != nil) {
        dispatch_source_cancel(self.heartbeatTimer);
        self.heartbeatTimer = nil;
    }
}

- (void)sendHeartbeat {
    if (self.webSocketTask == nil) {
        return;
    }
    NSURLSessionWebSocketMessage *message = [[NSURLSessionWebSocketMessage alloc] initWithString:@"{\"type\":\"heartbeat\"}"];
    [self.webSocketTask sendMessage:message completionHandler:^(NSError *error) {
        if (error != nil) {
            NSLog(@"[%@] heartbeat failed: %@", kRelayLogTag, error.localizedDescription);
        }
    }];
}

#pragma mark - Login

- (void)ensureLoggedIn {
    if (self.token.length > 0) {
        return;
    }
    [self login];
}

- (void)login {
    NSURL *url = [NSURL URLWithString:[NSString stringWithFormat:@"%@/api/client/login", self.baseURL]];
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url];
    request.HTTPMethod = @"POST";
    [request setValue:@"application/json; charset=UTF-8" forHTTPHeaderField:@"Content-Type"];
    NSDictionary *body = @{
        @"username": self.username ?: @"",
        @"password": self.password ?: @"",
        @"clientId": self.clientId ?: @"",
        @"group": self.group ?: @"",
        @"platform": self.platform ?: @"ios",
        @"maxInFlight": @(self.maxInFlight)
    };
    request.HTTPBody = [NSJSONSerialization dataWithJSONObject:body options:0 error:nil];

    dispatch_semaphore_t sem = dispatch_semaphore_create(0);
    __block NSData *responseData = nil;
    __block NSHTTPURLResponse *httpResponse = nil;
    __block NSError *requestError = nil;

    [[self.session dataTaskWithRequest:request completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
        responseData = data;
        if ([response isKindOfClass:[NSHTTPURLResponse class]]) {
            httpResponse = (NSHTTPURLResponse *)response;
        }
        requestError = error;
        dispatch_semaphore_signal(sem);
    }] resume];
    dispatch_semaphore_wait(sem, DISPATCH_TIME_FOREVER);

    if (requestError != nil) {
        @throw [NSException exceptionWithName:@"RelayClientError" reason:requestError.localizedDescription userInfo:nil];
    }
    if (httpResponse.statusCode >= 400) {
        @throw [NSException exceptionWithName:@"RelayClientError" reason:[NSString stringWithFormat:@"HTTP %ld", (long)httpResponse.statusCode] userInfo:nil];
    }

    id parsed = [NSJSONSerialization JSONObjectWithData:responseData options:0 error:nil];
    if (![parsed isKindOfClass:[NSDictionary class]]) {
        @throw [NSException exceptionWithName:@"RelayClientError" reason:@"invalid login response" userInfo:nil];
    }
    NSDictionary *responseBody = (NSDictionary *)parsed;
    NSString *token = [self stringValue:responseBody[@"token"]];
    if (token.length == 0) {
        @throw [NSException exceptionWithName:@"RelayClientError" reason:@"login succeeded but token is missing" userInfo:nil];
    }
    self.token = token;
    NSString *wsURL = [self stringValue:responseBody[@"wsUrl"]];
    self.wsURL = wsURL.length > 0 ? wsURL : [self buildWSURL];

    NSNumber *maxInFlight = responseBody[@"maxInFlight"];
    if ([maxInFlight isKindOfClass:[NSNumber class]] && maxInFlight.integerValue > 0) {
        self.maxInFlight = maxInFlight.integerValue;
    }
}

#pragma mark - Helpers

- (NSString *)normalizedBaseURL:(NSString *)value {
    NSString *normalized = [value stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
    if (normalized.length == 0) {
        @throw [NSException exceptionWithName:@"RelayClientError" reason:@"baseURL can not be empty" userInfo:nil];
    }
    if (![normalized hasPrefix:@"http://"] && ![normalized hasPrefix:@"https://"]) {
        normalized = [@"http://" stringByAppendingString:normalized];
    }
    while ([normalized hasSuffix:@"/"]) {
        normalized = [normalized substringToIndex:normalized.length - 1];
    }
    return normalized;
}

- (NSString *)buildWSURL {
    NSString *wsBase = self.baseURL;
    if ([wsBase hasPrefix:@"https://"]) {
        wsBase = [@"wss://" stringByAppendingString:[wsBase substringFromIndex:8]];
    } else if ([wsBase hasPrefix:@"http://"]) {
        wsBase = [@"ws://" stringByAppendingString:[wsBase substringFromIndex:7]];
    }
    NSString *encodedToken = [self urlEncode:self.token ?: @""];
    return [NSString stringWithFormat:@"%@/api/client/ws?token=%@", wsBase, encodedToken];
}

- (NSString *)urlEncode:(NSString *)value {
    NSMutableString *result = [NSMutableString string];
    const char *bytes = value.UTF8String;
    for (const char *cursor = bytes; cursor != NULL && *cursor != '\0'; cursor++) {
        unsigned char c = (unsigned char)*cursor;
        BOOL safe = (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || c == '-' || c == '_' || c == '.' || c == '~';
        if (safe) {
            [result appendFormat:@"%c", c];
        } else {
            [result appendFormat:@"%%%02X", c];
        }
    }
    return result;
}

- (NSTimeInterval)retryDelayForAttempt:(NSInteger)attempt {
    NSInteger safeAttempt = MAX(0, MIN(attempt, 6));
    NSTimeInterval capped = kBaseRetryDelay * pow(2.0, safeAttempt);
    capped = MIN(capped, kMaxRetryDelay);
    if (capped <= kBaseRetryDelay) {
        return kBaseRetryDelay;
    }
    return kBaseRetryDelay + ((double)arc4random_uniform((uint32_t)((capped - kBaseRetryDelay) * 1000.0)) / 1000.0);
}

- (NSString *)stringValue:(id)value {
    if (value == nil || value == [NSNull null]) {
        return @"";
    }
    if ([value isKindOfClass:[NSString class]]) {
        return (NSString *)value;
    }
    return [NSString stringWithFormat:@"%@", value];
}

@end
